feat(progress-bar): add inline variant

kevinsperrine opened the pull request:
Closes #3339 

**Summary**

- adds `inline` prop to ProgressBar to hide label/icon/helperText and show the value and unit inlined with the bar

**Change List (commits, features, bugs, etc)**

- add `inline` prop
- update docs
- add test
- update snapshots
- cleanup code

**Acceptance Test (how to verify the PR)**

- go to the [ProgressBar story](https://deploy-preview-3347--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-progressbar--default)
- turn on the inline prop
- confirm label and helperText disappear and value is inlined beside the bar with 8px of space in between
- repeat with icons/thresholds
- check RTL

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here

<!-- For reviewers: do not remove -->

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

bjornalm commented:
This looks good, just some edge case value scenarios we might wanna consider.

Not everything below is related to this PR  but I did some monkey testing with insane values and got this moving bar and a null value
![image](https://user-images.githubusercontent.com/5167091/155979692-704f990f-7316-4779-9031-873f6a1af1e0.png)
when using large enough numbers

The bar can completely disappear if the value or unit is large enough. Also managed to get a line break if the the unit has a space
![image](https://user-images.githubusercontent.com/5167091/155980258-b35064e6-0b05-4b74-bba8-044e88ebc21f.png)

And the text can overflow
![image](https://user-images.githubusercontent.com/5167091/155980436-dacbfe73-c9c6-4a48-8edf-4b5d63096116.png)

Overflowing for non inline version
![image](https://user-images.githubusercontent.com/5167091/155980986-46198e4f-5789-4ba0-aefd-039d209dfafc.png)

bjornalm commented on packages/react/src/components/ProgressBar/ProgressBar.jsx:
```diff
@@ -99,6 +103,72 @@ const ProgressBar = ({
+    >
+      <CarbonProgressBar
+        label={label}
+        helperText={!inline ? helperText : undefined}
```
We know this is inline right?

kevinsperrine replied:
Fixed.

bjornalm commented on packages/react/src/components/ProgressBar/ProgressBar.jsx:
```diff
@@ -99,6 +103,72 @@ const ProgressBar = ({
+      <CarbonProgressBar
+        label={label}
+        helperText={!inline ? helperText : undefined}
+        hideLabel={hideLabel || inline}
```
Same here, we can just always hide the label

kevinsperrine replied:
Fixed.

bjornalm commented on packages/react/src/components/ProgressBar/ProgressBar.jsx:
```diff
@@ -114,38 +184,20 @@ const ProgressBar = ({
+          {renderProgressBarLabel()}
+          <CarbonProgressBar
+            label={label}
+            helperText={!inline ? helperText : undefined}
```
Here we know that it is not inline

kevinsperrine replied:
Fixed.

bjornalm commented on packages/react/src/components/ProgressBar/ProgressBar.jsx:
```diff
@@ -114,38 +184,20 @@ const ProgressBar = ({
+          <CarbonProgressBar
+            label={label}
+            helperText={!inline ? helperText : undefined}
+            hideLabel={hideLabel || inline}
```
and here we can just use `hideLabel={hideLabel}`

bjornalm commented on packages/react/src/components/ProgressBar/ProgressBar.mdx:
```diff
@@ -92,6 +93,8 @@ Thresholds can be given to change the color of the bar (and icon) at various val
 | i18n                    | shape                               |                     |                                                                                                                             |
 | i18n.iconLabel          | string                              | 'Progress bar icon' | aria-label used on the icon                                                                                                 |
 
+- Denotes required props
```
```suggestion
\*) Denotes required props
```

kevinsperrine replied:
Fixed.

kevinsperrine commented:
> This looks good, just some edge case value scenarios we might wanna consider.
> 
> Not everything below is related to this PR but I did some monkey testing with insane values and got this moving bar and a null value ![image](https://user-images.githubusercontent.com/5167091/155979692-704f990f-7316-4779-9031-873f6a1af1e0.png) when using large enough numbers
> 
> The bar can completely disappear if the value or unit is large enough. Also managed to get a line break if the the unit has a space ![image](https://user-images.githubusercontent.com/5167091/155980258-b35064e6-0b05-4b74-bba8-044e88ebc21f.png)
> 
> And the text can overflow ![image](https://user-images.githubusercontent.com/5167091/155980436-dacbfe73-c9c6-4a48-8edf-4b5d63096116.png)
> 
> Overflowing for non inline version ![image](https://user-images.githubusercontent.com/5167091/155980986-46198e4f-5789-4ba0-aefd-039d209dfafc.png)

This should all be handled now. The moving bar is standard carbon behavior if no value is given. I hide the value/unit if it's not a number, and both the unit and the label should truncate when they are very long.

bjornalm commented:
Yes, it is working. But it raises the question, how long is very long in this context? It now truncates if I wanna show 450/500.
![image](https://user-images.githubusercontent.com/5167091/156039520-54cdd328-defb-45fd-8fff-afee4d710ecc.png)

And if I would like to show value "1000" and unit "MiB" it also truncates, i.e. anything over 3 digit unit and 4 digit values
![image](https://user-images.githubusercontent.com/5167091/156040418-507f4a21-efa4-4b2f-baf2-37c7c90cd5cf.png)

bjornalm reviewed (approved):
Looks good. Just the question regarding when to truncate... I guess that is a question for @dianatran18

kevinsperrine commented:
> Looks good. Just the question regarding when to truncate... I guess that is a question for @dianatran18

Yeah, I don't know. We're pretty limited, because we're wrapping carbon, and have to use absolute positioned elements. It would be better if we didn't have to set fixed limits instead of them shifting dynamically.

dianatran18 commented:
I might have to look into this further, but generally the text on the right of the bar should only be a percentage value, which is a max of 3 numbers and the `%`. I noted that the progress bar width would be dynamic based on the use case, but are we limited to a certain width?

Also for the label on the stacked variant, I would have the text truncate when it hits 16px from the percentage value or status icon (if there is one). These are the specs for that. You can see the bounding box for the label and where it would stop in both use cases.

We also have usage guidance stating that "The label must be concise, not exceeding the length of the progress bar."


![image](https://user-images.githubusercontent.com/43549567/156041618-394981bd-3a13-44fc-952f-29ec7f2d7a29.png)

kevinsperrine commented:
Thanks, @dianatran18 and @bjornalm. I've adjusted the widths and padding so that both the default and inline versions can show `400/500` without truncating, but more than that will truncate. I've also adjusted the padding on the normal version to always have 16px between either the icon (or the text if no icon is given) and that's the width at which the truncation of the label will start. Seems the best middle ground.

davidicus commented on packages/react/src/components/ProgressBar/ProgressBar.jsx:
```diff
@@ -98,13 +102,74 @@ const ProgressBar = ({
+   *
+   * @returns JSXElement
+   */
+  const renderValueAndUnit = () => (
```
Why do we call these as functions rather than use the JSX?

kevinsperrine replied:
Just habit to create a render function. No reason it can't be jsx in this case. Updated and pushed.
